File: frameworks/motia/motia-py/packages/motia/src/motia/state.py
"""State management for Motia framework."""

from __future__ import annotations

from typing import Any

from .iii import get_instance
from .tracing import operation_span, record_exception, set_span_ok

_list = list  # module-level alias; StateManager.list() shadows the builtin inside the class


class StateManager:
    """Internal state manager using state SDK calls."""

    def get(self, scope: str, key: str) -> Any | None:
        """Get a value from the state."""
        with operation_span(
            "state::get",
            **{"motia.state.scope": scope, "motia.state.key": key},
        ) as span:
            try:
                result = get_instance().trigger(
                    {
                        "function_id": "state::get",
                        "payload": {"scope": scope, "key": key},
                    }
                )
                set_span_ok(span)
                return result
            except Exception as exc:
                record_exception(span, exc)
                raise

    def set(self, scope: str, key: str, value: Any) -> Any:
        """Set a value in the state."""
        with operation_span(
            "state::set",
            **{"motia.state.scope": scope, "motia.state.key": key},
        ) as span:
            try:
                result = get_instance().trigger(
                    {
                        "function_id": "state::set",
                        "payload": {"scope": scope, "key": key, "value": value},
                    }
                )
                set_span_ok(span)
                return result
            except Exception as exc:
                record_exception(span, exc)
                raise

    def update(self, scope: str, key: str, ops: list[dict[str, Any]]) -> Any:
        """Update a value in the state using update operations."""
        with operation_span(
            "state::update",
            **{"motia.state.scope": scope, "motia.state.key": key},
        ) as span:
            try:
                result = get_instance().trigger(
                    {
                        "function_id": "state::update",
                        "payload": {"scope": scope, "key": key, "ops": ops},
                    }
                )
                set_span_ok(span)
                return result
            except Exception as exc:
                record_exception(span, exc)
                raise

    def delete(self, scope: str, key: str) -> Any | None:
        """Delete a value from the state."""
        with operation_span(
            "state::delete",
            **{"motia.state.scope": scope, "motia.state.key": key},
        ) as span:
            try:
                result = get_instance().trigger(
                    {
                        "function_id": "state::delete",
                        "payload": {"scope": scope, "key": key},
                    }
                )
                set_span_ok(span)
                return result
            except Exception as exc:
                record_exception(span, exc)
                raise

    def list(self, scope: str) -> list[Any]:
        """List all values in a scope."""
        with operation_span(
            "state::list",
            **{"motia.state.scope": scope},
        ) as span:
            try:
                items: list[Any] = get_instance().trigger(
                    {
                        "function_id": "state::list",
                        "payload": {"scope": scope},
                    }
                )
                set_span_ok(span)
                return items
            except Exception as exc:
                record_exception(span, exc)
                raise

    def list_groups(self) -> _list[str]:
        """List all scope IDs."""
        with operation_span("state::list_groups") as span:
            try:
                groups: _list[str] = get_instance().trigger({"function_id": "state::list_groups", "payload": {}})
                set_span_ok(span)
                return groups
            except Exception as exc:
                record_exception(span, exc)
                raise

    def clear(self, scope: str) -> None:
        """Clear all values in a scope."""
        with operation_span(
            "state::clear",
            **{"motia.state.scope": scope},
        ) as span:
            try:
                items = self.list(scope)
                for item in items:
                    if isinstance(item, dict) and "id" in item:
                        self.delete(scope, item["id"])
                set_span_ok(span)
            except Exception as exc:
                record_exception(span, exc)
                raise


stateManager = StateManager()

File: frameworks/motia/motia-js/packages/motia/src/new/state.ts
import type { StreamSetResult, UpdateOp } from 'iii-sdk/stream'
import { SpanStatusCode, withSpan } from 'iii-sdk/telemetry'
import type { InternalStateManager } from '../types'
import { getInstance } from './iii'

export class StateManager implements InternalStateManager {
  async get<T>(scope: string, key: string): Promise<T | null> {
    return withSpan('state::get', {}, async (span) => {
      span.setAttribute('motia.state.scope', scope)
      span.setAttribute('motia.state.key', key)
      try {
        return (await getInstance().trigger({ function_id: 'state::get', payload: { scope, key } })) as T | null
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }

  async set<T>(scope: string, key: string, value: T): Promise<StreamSetResult<T> | null> {
    return withSpan('state::set', {}, async (span) => {
      span.setAttribute('motia.state.scope', scope)
      span.setAttribute('motia.state.key', key)
      try {
        return (await getInstance().trigger({
          function_id: 'state::set',
          payload: { scope, key, value },
        })) as StreamSetResult<T> | null
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }

  async delete<T>(scope: string, key: string): Promise<T | null> {
    return withSpan('state::delete', {}, async (span) => {
      span.setAttribute('motia.state.scope', scope)
      span.setAttribute('motia.state.key', key)
      try {
        return (await getInstance().trigger({ function_id: 'state::delete', payload: { scope, key } })) as T | null
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }

  async list<T>(scope: string): Promise<T[]> {
    return withSpan('state::list', {}, async (span) => {
      span.setAttribute('motia.state.scope', scope)
      try {
        return (await getInstance().trigger({ function_id: 'state::list', payload: { scope } })) as T[]
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }

  async listGroups(): Promise<string[]> {
    return withSpan('state::list_groups', {}, async (span) => {
      try {
        return (await getInstance().trigger({ function_id: 'state::list_groups', payload: {} })) as string[]
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }

  async update<T>(scope: string, key: string, ops: UpdateOp[]): Promise<StreamSetResult<T> | null> {
    return withSpan('state::update', {}, async (span) => {
      span.setAttribute('motia.state.scope', scope)
      span.setAttribute('motia.state.key', key)
      try {
        return (await getInstance().trigger({ function_id: 'state::update', payload: { scope, key, ops } })) as StreamSetResult<T> | null
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }

  async clear(scope: string): Promise<void> {
    return withSpan('state::clear', {}, async (span) => {
      span.setAttribute('motia.state.scope', scope)
      try {
        const items = await this.list<{ id: string }>(scope)

        for (const item of items) {
          await this.delete(scope, item.id)
        }
      } catch (err) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) })
        span.recordException(err as Error)
        throw err
      }
    })
  }
}

export const stateManager = new StateManager()
